### PR TITLE
Pass key to tasks.

### LINKED ISSUE #1969: When network interface isn't specified, task processes fail to connect to driver.
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): N/A
2. Framework version: N/A
3. Horovod version: 0.19.2
4. MPI version: N/A (Using Gloo Controller)
5. CUDA version: 10.0
6. NCCL version: 2.6.4
7. Python version: 3.6.10
8. OS and version: Ubuntu 18.04
9. GCC version: 4.8

**Bug report:**

When running multi-machine and not specifying a network interface, task processes fail to connect to driver. Note: using a GLOO controller.

Stack trace:

```
root@ip-172-31-37-52:/# horovodrun -np 2 -H localhost:1,172.31.35.37:1 -p 12345 --verbose ls
Filtering local host names.
Remote host found: 172.31.35.37
Checking ssh on all remote hosts.
SSH was successful into all the remote hosts.
Testing interfaces on all the hosts.
Launched horovod server.
Attempted to launch horovod task servers.
Waiting for the hosts to acknowledge.
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 37436)
Exception happened during processing of request from ('172.31.37.52', 52632)
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 52636)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
Traceback (most recent call last):
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 55358)
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 52642)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
Exception happened during processing of request from ('127.0.0.1', 37442)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
struct.error: unpack requires a buffer of 4 bytes
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 55364)
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 37448)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
----------------------------------------
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 55368)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.35.37', 39790)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.35.37', 39796)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
----------------------------------------
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
Exception happened during processing of request from ('172.31.35.37', 39798)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
Launching horovod task function was not successful:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/task_fn.py", line 67, in <module>
    _task_fn(index, driver_addresses, settings)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/task_fn.py", line 27, in _task_fn
    driver_addresses, settings.key, settings.verbose)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/driver/driver_service.py", line 44, in __init__
    match_intf=match_intf)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/service/driver_service.py", line 159, in __init__
    match_intf=match_intf)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 172, in __init__
    'Linux.'.format(service_name=service_name, addresses=addresses))
horovod.run.common.util.network.NoValidAddressesFound: Horovod was unable to connect to horovod driver service on any of the following addresses: {'lo': [('127.0.0.1', 4548)], 'ens3': [('172.31.37.52', 4548)], 'docker0': [('172.17.0.1', 4548)]}.

One possible cause of this problem is that horovod currently requires every host to have at least one routable network interface with the same name across all of the hosts. You can run "ifconfig -a" on every host and check for the common routable interface. To fix the problem, you can rename interfaces on Linux.
```

When a network interface is specified, everything works as expected:
```
root@ip-172-31-37-52:/# horovodrun -np 2 -H localhost:1,172.31.35.37:1 -p 12345 --network-interface ens3 ls
Wed May 20 19:01:46 2020[0]<stdout>:bin
Wed May 20 19:01:46 2020[0]<stdout>:boot
.....
```

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
Previously the driver would omit serializing the secret key, making it impossible for tasks to
connect to the driver when network interface is not specified.

Fixes  #1969.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by tgaddair** [comment]
Thanks for the PR, @aaron276h.  @EnricoMi, can you review since this was code you recently worked on?

**Comment by EnricoMi** [comment]
@aaron276h let me check if we can avoid adding the key back to serialisation.

**Comment by tgaddair** [comment]
@EnricoMi seems there are more people reporting this issue.  We may want to go ahead and land this and release a hotfix (v0.19.3) to unblock users, then add back the secret key removal once a more complete fix is found.

@aaron276h can you disable `RunTests.test_settings_dump_drops_key` for now to make tests pass?

**Comment by aaron276h** [comment]
@tgaddair Updated the pr with the test disabled. 

One medium term solution could be to remove the secret key from the serialization of `Settings` only when running with spark, since that is the only scenario when it is being passed correctly to the tasks.

File: horovod/run/common/util/settings.py [hunk 1]
```diff
@@ -57,11 +57,8 @@ def __init__(self, num_proc=None, verbose=0, ssh_port=None, extra_mpi_args=None,
         self.nics = nics
         self.elastic = elastic
         
-    # we do not serialize the key, as it is too risky that it could leak unintentionally
     def __getstate__(self):
```
Comment (EnricoMi) [inline horovod/run/common/util/settings.py:60]: you can remove the `__getstate__` method entirely

**Comment by EnricoMi** [comment]
> @tgaddair Updated the pr with the test disabled.
> 
> One medium term solution could be to remove the secret key from the serialization of `Settings` only when running with spark, since that is the only scenario when it is being passed correctly to the tasks.

Done in https://github.com/horovod/horovod/pull/1974. @tgaddair please consider that PR to also go int v0.19.3.

Medium term solution should be to remove the key also for non-Spark MPI (should be trivial) and Gloo (will be hard).

**Comment by aaron276h** [comment]
Closing since #1974 fixed the underlying issue.

**Pull request closed without merging**